Src/FluentAssertions/Formatting/DefaultValueFormatter.cs

```diff
@@ -72,10 +80,32 @@ private static bool HasDefaultToStringImplementation(object value)
     private void WriteTypeAndMemberValues(object obj, FormattedObjectGraph formattedGraph, FormatChild formatChild)
     {
         Type type = obj.GetType();
+        WriteTypeName(formattedGraph, type);
+        WriteTypeValue(obj, formattedGraph, formatChild, type);
+    }
+
+    private void WriteTypeName(FormattedObjectGraph formattedGraph, Type type)
+    {
         formattedGraph.AddLine(TypeDisplayName(type));
-        formattedGraph.AddLine("{");
+    }
```

Comment:
The nested anonymous type `When_the_object_is_an_anonymous_type_it_should_show_the_properties_recursively` is formatted as this.
Note the empty lines because of `TypeDisplayName` returning an empty string.

```
Expected stuff to be 
{
	Children = {10, 20, 30, 40}, 
	SingleChild = 
	
	{
		ChildId = 4
	}
}, but found 
{
	Children = {10, 20, 30, 40}, 
	Description = "absent", 
	SingleChild = 
	
	{
		ChildId = 4
	}
}.
```

Here are two different ways that, _for this example_, both produce
```
Expected stuff to be 
{
	Children = {10, 20, 30, 40}, 
	SingleChild = 
	{
		ChildId = 4
	}
}, but found 
{
	Children = {10, 20, 30, 40}, 
	Description = "absent", 
	SingleChild = 
	{
		ChildId = 4
	}
}.
```

1)
```cs
private void WriteTypeName(FormattedObjectGraph formattedGraph, Type type)
{
    if (type.HasTypeName())
    {
        formattedGraph.AddLine(TypeDisplayName(type));
    }
    else if (formattedGraph.LineCount == 0)
    {
        formattedGraph.AddLine(string.Empty);
    }
}

protected virtual string TypeDisplayName(Type type) => type.FullName;
```

2)
```cs
private void WriteTypeName(FormattedObjectGraph formattedGraph, Type type)
{
    var typeName = TypeDisplayName(type);
    if (formattedGraph.LineCount == 0 || !string.IsNullOrEmpty(typeName))
    {
        formattedGraph.AddLine(typeName);
    }
}

protected virtual string TypeDisplayName(Type type) => type.HasTypeName() ? type.FullName : string.Empty;
```

I haven't thought much about if 1) or 2) is best or a third option is even better.

Reply:
Good spot!  After going down a rabbit hole for a while I discovered that `.WithMessage` wildcards ignore newlines.  I also found that I was putting the type name on a new line as well, which was not intentional.

I've reworked some tests so that they are explicit about where the newlines should go, and ended up with the following.  I pushed the responsibility for determining whether or not to add a newline into `FormattedObjectGraph` to keep use of its internals local:

```c#
    private void WriteTypeName(FormattedObjectGraph formattedGraph, Type type)
    {
        var typeName = TypeDisplayName(type);
        formattedGraph.AddFragment(typeName);
    }

    private void WriteTypeValue(object obj, FormattedObjectGraph formattedGraph, FormatChild formatChild, Type type)
    {
        MemberInfo[] members = GetMembers(type);
        if (members.Length == 0)
        {
            WriteEmptyType(formattedGraph);
        }
        else
        {
            WriteTypeValue(obj, formattedGraph, formatChild, members);
        }
    }

    private static void WriteEmptyType(FormattedObjectGraph formattedGraph)
    {
        formattedGraph.AddFragment("{ }");
    }

    private static void WriteTypeValue(object obj, FormattedObjectGraph formattedGraph, FormatChild formatChild, MemberInfo[] members)
    {
        formattedGraph.EnsureNewLine();
        formattedGraph.AddLine("{");
        WriteMemberValues(obj, members, formattedGraph, formatChild);
        formattedGraph.AddFragmentOnNewLine("}");
    }
```

Output for defined types (`When_the_object_is_a_user_defined_type_it_should_show_the_name_on_the_initial_line`):

    Expected stuff to be <null>, but found FluentAssertions.Specs.Formatting.FormatterSpecs+StuffRecord
    {
        RecordChildren = {10, 20, 30, 40}, 
        RecordDescription = "description", 
        RecordId = 42, 
        SingleChild = FluentAssertions.Specs.Formatting.FormatterSpecs+ChildRecord
        {
            ChildRecordId = 24
        }
    }.

Output for anonymous types (`When_the_object_is_an_anonymous_type_it_should_show_the_properties_recursively`):

    Expected stuff to be <null>, but found 
    {
        Children = {1, 2, 3, 4}, 
        Description = "absent", 
        SingleChild = 
        {
            ChildId = 8
        }
    }.

Output for empty anonymous type (`When_the_object_is_an_empty_anonymous_type_it_should_show_braces_on_the_same_line`):

    Expected stuff to be <null>, but found { }.